File: solutions/fortran/nth-prime/1/nth_prime.f90
module nth_prime
  implicit none
contains

  ! get nth prime
  integer function prime(n)
    integer, intent(in) :: n
    integer :: i, candidate, count
    logical :: is_prime
    
    if (n <= 0) then
      prime = -1
      return
    end if
    
    if (n == 1) then
      prime = 2
      return
    end if
    
    count = 1  ! we already have prime #1 (2)
    candidate = 3  ! start checking from 3
    
    do while (count < n)
      is_prime = .true.
      
      ! Check if candidate is prime
      ! Only check up to sqrt(candidate)
      do i = 2, int(sqrt(real(candidate)))
        if (mod(candidate, i) == 0) then
          is_prime = .false.
          exit
        end if
      end do
      
      if (is_prime) then
        count = count + 1
        if (count == n) then
          prime = candidate
          return
        end if
      end if
      
      candidate = candidate + 2  ! check only odd numbers (even > 2 are not prime)
    end do
    
    prime = candidate
  end function prime

end module nth_prime

File: solutions/cpp/nth-prime/1/nth_prime.cpp
// nth_prime.cpp
#include "nth_prime.h"
#include <stdexcept>

namespace nth_prime {

int nth(int n) {
    if (n <= 0) {
        throw std::domain_error("n must be a positive integer");
    }

    int count = 0;
    int candidate = 2;

    while (true) {
        bool is_prime = true;
        // Check divisibility up to sqrt(candidate)
        for (int i = 2; i * i <= candidate; ++i) {
            if (candidate % i == 0) {
                is_prime = false;
                break;
            }
        }
        if (is_prime) {
            ++count;
            if (count == n) {
                return candidate;
            }
        }
        // Increment candidate; after 2 we can skip even numbers
        ++candidate;
    }
}

}  // namespace nth_prime